### DISCUSSION TITLE
vm.prank( ) cannot understand .

### DISCUSSION THREAD

**Discussion by Debangshu97** [Cyfrin/foundry-full-course-cu#164] [Q&A] [answered]
Can someone explain the use of vm.prank with an example.

**Comment by ghost** [comment, marked as answer]
Before speaking about `prank`, let me explain a little bit about what I've learned about Foundry.

In foundry everything is a conctract -- scripts, tests, libraries, imports, etc..., due to this reason Foundry has set up a serie of default addresses that acts as the `msg.sender` to execute these files and the functions in them.

When we execute a test, is not our private key that is being used, in fact is a key from Foundry.

```javascript
function testConsoleCaller() public view {
    console.log(msg.sender); // is not you!
}
```

This behaviour could be problematic depending on the test, for example the `withdraw` function from the `FundMe` contract has a modifier that only allows the owner to withdraw the funds, but by default the `msg.sender` in the test file is a default key from Foundry not you.

```javascript
function testCanWithdrawTheFund() public {
    fundMe.withdraw(); // the test fails because the default address that execute the file is not the owner
}
```

And for cases like this one `prank` comes into play. `prank` allows us to set the `msg.sender` for the **next call**. 

```javascript
function testCanWithdrawTheFund() public {
    vm.prank(USER); // your key will be used to withdraw the funds
    fundMe.withdraw();
}
```

`prank` only works one time with the first transaction it sees:

```javascript
function testCanFundAndWithdraw() public {
    vm.prank(USER); // your key will be used to fund the contract
    fundMe.fund{value: 1 ether}(); // prank only works in this transaction and then it's over
    fundMe.withdraw(); // the `msg.sender` is again the key from Foundry
}
```
If you want to bundle transactions use `startPrank` and `stopPrank`:

```solidity
function testCanWithdrawTheFund() public {
    vm.startPrank(USER); // your key will be used for all the next transactions until you call `stopPrank`
    fundMe.fund{value: 1 ether}();
    fundMe.withdraw();
    vm.stopPrank();
}
```

This is a recopilation of the behaviours I have noticed from `prank`.

**Reply by cromewar** [reply]
Amazing answer @pacelliv

**Comment by Debangshu97** [comment]
thanks man!
